pkg/provider/scaleway/client.go

```diff
@@ -0,0 +1,444 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package scaleway
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	smapi "github.com/scaleway/scaleway-sdk-go/api/secret/v1alpha1"
+	"github.com/scaleway/scaleway-sdk-go/scw"
+	"github.com/tidwall/gjson"
+
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	"github.com/external-secrets/external-secrets/pkg/find"
+)
+
+var errNoSecretForName = errors.New("no secret for this name")
+
+type client struct {
+	api       secretAPI
+	projectID string
+	cache     cache
+}
+
+const (
+	refTypeName = "name"
+	refTypeID   = "id"
+)
+
+type scwSecretRef struct {
+	RefType string
+	Value   string
+}
+
+func (r scwSecretRef) String() string {
+	return fmt.Sprintf("%s:%s", r.RefType, r.Value)
+}
+
+func decodeScwSecretRef(key string) (*scwSecretRef, error) {
+	sepIndex := strings.IndexRune(key, ':')
+	if sepIndex < 0 {
+		return nil, fmt.Errorf("invalid secret reference: missing colon ':'")
+	}
+
+	return &scwSecretRef{
+		RefType: key[:sepIndex],
+		Value:   key[sepIndex+1:],
+	}, nil
+}
+
+func (c *client) getSecretByName(ctx context.Context, name string) (*smapi.Secret, error) {
+	request := smapi.GetSecretByNameRequest{
+		SecretName: name,
+	}
+
+	response, err := c.api.GetSecretByName(&request, scw.WithContext(ctx))
+	if err != nil {
+		//nolint:errorlint
+		if _, isErrNotFound := err.(*scw.ResourceNotFoundError); isErrNotFound {
+			return nil, errNoSecretForName
+		}
+		return nil, err
+	}
+
+	return response, nil
+}
+
+func (c *client) GetSecret(ctx context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	scwRef, err := decodeScwSecretRef(ref.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	versionSpec := "latest_enabled"
+	if ref.Version != "" {
+		versionSpec = ref.Version
+	}
+
+	value, err := c.accessSecretVersion(ctx, scwRef, versionSpec)
+	if err != nil {
+		//nolint:errorlint
+		if _, isNotFoundErr := err.(*scw.ResourceNotFoundError); isNotFoundErr {
+			return nil, esv1beta1.NoSecretError{}
+		}
+		return nil, err
+	}
+
+	if ref.Property != "" {
+		extracted, err := extractJSONProperty(value, ref.Property)
+		if err != nil {
+			return nil, err
+		}
+
+		value = extracted
+	}
+
+	return value, nil
+}
+
+func (c *client) PushSecret(ctx context.Context, value []byte, remoteRef esv1beta1.PushRemoteRef) error {
+	scwRef, err := decodeScwSecretRef(remoteRef.GetRemoteKey())
```

Comment:
Do we really have to use `name:` here? if this is the only way to push secrets then i guess it can be omitted for brevity.

Reply:
Wouldn't it be confusing to the user to have two different syntaxes for remote references?
We could default to `name:` everywhere, but it would be less clear that the two means of referencing secrets exist (the id is preferred when possible because it has no unicity issues).
What do you think?

Reply:
Fair points!
Sorry for the wall of text, i just wanted to dump my thoughts here. Naming is hard :joy:  :sweat_smile: 

(1) fix name collision upstream (???)
IMHO both options to reference secrets in Scaleway SM are suboptimal: 
1. `id` is hard to work with: hard to read, risk of human error
2. `name` is ambiguous, hence (in a strict sense) useless

My gut feeling tells me that this is a API design fault and i guess we're not the only ones suffering from that. I reckon there will/is a terraform provider out there that uses this API that suffers from that problem. Given that the API is in beta this is still fixable i hope. Honestly i don't have the time nor the network to start a discussion with this vendor. Maybe you have or already know the answer that this is a WONTFIX :thinking:.

(2) remoteRef.key syntax
Moving forward, i agree, we should be explicit with the syntax for the remote reference.
For context: Azure KeyVault has different object types you can store: `key`, `cert`, `secret` and their remote reference contains this type in the `remoteRef.key` and uses this schema: `secret/{name}`, `key/{name}` etc.
Having and `id:` or `name:` prefix looks a bit odd (reading it as yaml string) and a bit inconsistent with the Azure KeyVault implementation. 
Of course, semantically these are two different things: `object-type` vs `reference-type`. Hence a different syntax is appropriate.

Long-term it may be worth looking into standardising this into a RFC3986 URI compatible scheme, but that's a whole other story.

Simply put: yep, let's leave it as you have suggested :grin:

Reply:
I can confirm it will probably stay a WONTFIX... As you say naming is hard :)